### PR TITLE
feat: RPC - starknet_syncing

### PR DESCRIPTION
**Pull request by zediogoviana**
RPC Endpoint that returns an object about the sync status, or false if the node is not syncing.

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

Creating a new RPC Endpoint.

Issue Number: #205 

# What is the new behavior?

This new endpoint looks for `starting_block`, `current_block` and `highest_block`, provided by the `sync_service`. With each block, we get the corresponding Starknet block and return its numbers and hashes respectively.

If the node is not syncing, then it returns false.

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

# Other information

### CODE REVIEW

File: crates/client/rpc/src/lib.rs [hunk 1]
```diff
@@ -11,31 +11,40 @@ use std::sync::Arc;
...
     overrides: Arc<OverrideHandle<B>>,
+    sync_service: Arc<SyncingService<B>>,
+    starting_block: <<B>::Header as HeaderT>::Number,
     _marker: PhantomData<(B, BE)>,
 }
 
 impl<B: BlockT, BE, C> Starknet<B, BE, C> {
-    pub fn new(client: Arc<C>, backend: Arc<mc_db::Backend<B>>, overrides: Arc<OverrideHandle<B>>) -> Self {
-        Self { client, backend, overrides, _marker: PhantomData }
+    pub fn new(
```
Comment (abdelhamidbakhta) [inline crates/client/rpc/src/lib.rs:40]: please add rust doc

File: crates/client/rpc/src/lib.rs [hunk 2]
```diff
@@ -196,6 +206,58 @@ where
             Err(e) => Err(e.into()),
         }
     }
+
+    async fn syncing(&self) -> RpcResult<Syncing> {
```
Comment (abdelhamidbakhta) [inline crates/client/rpc/src/lib.rs:210]: please add rust doc

Comment (abdelhamidbakhta) [reply crates/client/rpc/src/lib.rs:210]: please add comments in the function body to explain the steps

File: crates/client/rpc/src/madara_backend_client.rs [hunk 3]
```diff
@@ -31,3 +37,25 @@ where
     }
     false
 }
+
+pub fn starknet_block_from_substrate_hash<B: BlockT, C, BE>(
```
Comment (abdelhamidbakhta) [inline crates/client/rpc/src/madara_backend_client.rs:41]: please add rust doc

**Review by abdelhamidbakhta — CHANGES_REQUESTED** [review]
left some comments. also integration tests are missing for the new RPC

**Comment by EvolveArt** [comment]
Thanks a lot for this @zediogoviana ! Would it be possible to add some integration tests in typescript as well ?

**Comment by zediogoviana** [comment]
Yeah @EvolveArt , I'll do it. This week I've been attending ElixirConf in Lisbon, so time has been short to pick this up.

**Comment by zediogoviana** [comment]
After talking with @EvolveArt, I’ve been reading about Zombienet and how it could help with the missing test for the syncing RPC endpoint (the situation where the endpoint returns `false`). I’ve also searched on [substrate's node options](https://docs.substrate.io/reference/command-line-tools/node-template/) to see if there is anything related to syncing that I can toggle, but nothing seems to be what I want… I also found [this question](https://substrate.stackexchange.com/questions/6472/how-to-perform-manual-sync) that’s pretty much what I want, but the answer says the guy can’t do it, and he should try `--in-peers` and `--out-peers` to `0` but that’s already the case we have in the current test setup.

This way, @EvolveArt suggested moving ahead and opening an issue with these findings. So if everyone agrees, and when this gets merged, I'll do just that.

File: crates/node/src/rpc/starknet.rs [hunk 4]
```diff
@@ -12,6 +14,8 @@ pub struct StarknetDeps<C, B: BlockT> {
     pub madara_backend: Arc<Backend<B>>,
     /// Starknet data access overrides.
     pub overrides: Arc<OverrideHandle<B>>,
+    pub sync_service: Arc<SyncingService<B>>,
```
Comment (EvolveArt) [inline crates/node/src/rpc/starknet.rs:17]: rustdoc

File: crates/node/src/rpc/starknet.rs [hunk 5]
```diff
@@ -12,6 +14,8 @@ pub struct StarknetDeps<C, B: BlockT> {
...
+    pub starting_block: <<B>::Header as HeaderT>::Number,
```
Comment (EvolveArt) [inline crates/node/src/rpc/starknet.rs:18]: rustdoc

File: tests/tests/test-rpc/test-starknet-rpc.ts [hunk 6]
```diff
@@ -80,4 +80,16 @@ describeDevMadara("Starknet RPC", (context) => {
...
   });
+
+  it("syncing", async function () {
+    const status = await providerRPC.getSyncingStats();
+
+    expect(status["sync_status"]["starting_block_num"]).to.be.equal(0);
+    expect(status["sync_status"]["current_block_num"]).to.be.greaterThan(0);
+    expect(status["sync_status"]["highest_block_num"]).to.be.greaterThan(0);
+
+    expect(status["sync_status"]["starting_block_hash"]).to.contain("0xaf");
```
Comment (EvolveArt) [inline tests/tests/test-rpc/test-starknet-rpc.ts:91]: explain why these values in comments

**Review by EvolveArt — DISMISSED** [review]
left few comments

**Review by abdelhamidbakhta — APPROVED** [review]
lgtm

**Review by abdelhamidbakhta — APPROVED** [review]
lgtm

**Pull request merged**